src/x3d/x3dloadinternalstarling.pas

```diff
@@ -0,0 +1,611 @@
+  Shape: TShapeNode;
+  Tri: TTriangleSetNode;
+  Tex: TImageTextureNode;
+  Material: TUnlitMaterialNode;
+begin
+  Shape:= TShapeNode.Create;
+  Material := TUnlitMaterialNode.Create;
+  Material.EmissiveColor := Vector3(1, 1, 1);
+  Shape.Material := Material;
+
+  Tex := TImageTextureNode.Create;
+  Tex.FdUrl.Send(FImagePath);
+  Tex.RepeatS := false;
+  Tex.RepeatT := false;
+  Shape.Texture := Tex;
+
+  Tri := TTriangleSetNode.Create;
+  Tri.Solid := false;
+
+  FShapeCoord := TCoordinateNode.Create('coord');
+  FShapeCoord.SetPoint([
+      CoordArray[0],
+      CoordArray[1],
+      CoordArray[2],
+      CoordArray[3],
+      CoordArray[4],
+      CoordArray[5]]);
+
+  FShapeTexCoord := TTextureCoordinateNode.Create('texcoord');
+  FShapeTexCoord.SetPoint([
+       TexCoordArray[0],
+       TexCoordArray[1],
+       TexCoordArray[2],
+       TexCoordArray[3],
+       TexCoordArray[4],
+       TexCoordArray[5]]);
+
+  Tri.Coord := FShapeCoord;
+  Tri.TexCoord := FShapeTexCoord;
+  Shape.Geometry := Tri;
+
+  FRoot.AddChildren(Shape);
+end;
+
+procedure TStarlingTextureAtlasLoader.AddFrameCoords(
+    const CoordInterp: TCoordinateInterpolatorNode;
+    const TexCoordInterp: TCoordinateInterpolator2DNode);
+begin
+  CoordInterp.FdKeyValue.Items.AddRange(FCoordArray);
+  TexCoordInterp.FdKeyValue.Items.AddRange(FTexCoordArray);
+  { Repeat all keyValues, to avoid interpolating them smoothly between two keys }
+  CoordInterp.FdKeyValue.Items.AddRange(FCoordArray);
+  TexCoordInterp.FdKeyValue.Items.AddRange(FTexCoordArray);
+end;
+
+procedure TStarlingTextureAtlasLoader.AddAnimation(const FrameCount: Integer;
+  const TimeSensor: TTimeSensorNode;
+  const CoordInterp: TCoordinateInterpolatorNode;
+  const TexCoordInterp: TCoordinateInterpolator2DNode);
+var
+  I: Integer;
+  Key: Single;
+begin
+  { Set Cycle Interval becouse we know now frame count }
+  TimeSensor.CycleInterval := FrameCount / FFramesPerSecond;
+
+  { Generate list of keys. }
+  for I := 0 to FrameCount - 1 do
+  begin
+    Key := I / FrameCount;
+
+    CoordInterp.FdKey.Items.Add(Key);
+    TexCoordInterp.FdKey.Items.Add(Key);
+    if I > 0 then
+    begin
+      CoordInterp.FdKey.Items.Add(Key);
+      TexCoordInterp.FdKey.Items.Add(Key);
+    end;
+  end;
+
+  { This way, we have keys like
+    0 0.333 0.333 0.666 0.666 1
+    That is, all keys are repeated, except 0 and 1. }
+  CoordInterp.FdKey.Items.Add(1.0);
+  TexCoordInterp.FdKey.Items.Add(1.0);
+
+  { Add TimeSensor, CoordinateInterpolatorNode,
+    CoordinateInterpolator2DNode to Root node }
+  FRoot.AddChildren(TimeSensor);
+  FRoot.AddChildren(CoordInterp);
+  FRoot.AddChildren(TexCoordInterp);
+  AddRoutes(TimeSensor, CoordInterp, TexCoordInterp);
+end;
+
+procedure TStarlingTextureAtlasLoader.AddRoutes(
+  const TimeSensor: TTimeSensorNode;
+  const CoordInterp: TCoordinateInterpolatorNode;
+  const TexCoordInterp: TCoordinateInterpolator2DNode);
+var
+  R1, R2, R3, R4: TX3DRoute;
+begin
+  { Create routes. }
```

Comment:
I added a bit friendlier overload to add routes some time ago (I know, many loaders have not yet updated to using it).

Instead of

```
R1 := TX3DRoute.Create;
...
R1.SetSourceDirectly(TimeSensor.EventFraction_changed);
R1.SetDestinationDirectly(CoordInterp.EventSet_fraction);
...
FRoot.AddRoute(R1);
```

use 

```
FRoot.AddRoute(TimeSensor.EventFraction_changed, CoordInterp.EventSet_fraction);
```

This is shorter and usually easier to read.